### ISSUE TITLE
[typescript] tsserver no longer starts automatically when opening first file for a project

### ISSUE THREAD

**Issue by jupl** [syl20bnr/spacemacs#10662]
#### Description :octocat:
tsserver for tide/lsp does not start automatically.

#### Reproduction guide :beetle:
- Start Emacs with the TypeScript layer
- Open a `.ts` file in a TS project.

*Observed behaviour:* :eyes: :broken_heart:
Tide/LSP is not running. I need to do `M-x tide-setup` manually (for Tide), `revert-buffer`, or open another file in the project.

*Expected behaviour:* :heart: :smile:
Tide/LSP should automatically start as in previous versions. (at least for Tide)

#### System Info :computer:
- OS: darwin
- Emacs: 25.3.3
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. aa5bc0373)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(helm auto-completion emacs-lisp git lsp neotree syntax-checking typescript version-control)
```
- System configuration features: NOTIFY ACL LIBXML2 ZLIB TOOLKIT_SCROLL_BARS

**Comment by jupl** [comment]
Commit 59e0c56 (Add: lsp-typescript support) seemed to have broken it, as the commit just before that works as expected.

**Comment by jupl** [comment]
@ztlevi Can you verify this? Appears you were the author for this.

**Comment by ztlevi** [comment]
Could you try to `revert-buffer` after couple seconds just waiting for `tide` to set up properly?

I have the similar issue with the backend startup and it's due to the process is not async enough to enable the backend in Emacs after the backend is started. And a lot of time, I have to `revert-buffer` to make sure everything is running correctly.

I use `revert-buffer-no-confirm` cuz it's faster.
```
;; refresh buffer
(defun revert-buffer-no-confirm ()
  "Revert buffer without confirmation."
  (interactive)
  (save-buffer)
  (revert-buffer :ignore-auto :noconfirm))
```

**Comment by ztlevi** [comment]
It works fine on my machine... Including `company` and `jump-handler`.
![image](https://user-images.githubusercontent.com/16655096/39897213-333fb4f2-5466-11e8-81fb-fd556ebbd7b7.png)

**Comment by jupl** [comment]
Yes, `revert-buffer` does cause it to work again. This still is defect that needs to be fixed in Spacemacs. It's counterintuitive for one to do this every time I open a project or have to write custom code for this in my config file when it was working before. Also tried this with a fresh install.

**Comment by jupl** [comment]
And yes, I saw the LSP support and I'll mess around with that in my own time (as I need to install more dependencies and verify it still works on my machine) which I am excited for. In the meantime, I do rely on Tide for day to day to work and it has been working fine for me.

**Comment by jupl** [comment]
Tested locally LSP to mess around. Unfortunately I still need to do `revert-buffer` if I use LSP as well, meaning the TS layer is technically not working properly.

**Comment by ztlevi** [comment]
@jupl I found it's the recursive hooks problem. Not the async one... As long as this [PR](https://github.com/syl20bnr/spacemacs/pull/10664) get passed, you should be good to use OOB. Or you can apply the changes to your local repo as a temporary fix.

**Comment by ztlevi** [comment]
@jupl The problem is Purcell excludes some necessary files of `lsp-javascript-typescript` from mepla which will cause cannot find `lsp-javascript-flow` or `lsp-typescript` issue.
https://github.com/melpa/melpa/commit/97b6ddeb34297aacfefdd4c227ad520b70a12bc6

Try to replace [this line](https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/javascript/packages.el#L33) with `        (lsp-javascript-typescript :requires lsp-mode :location (recipe :fetcher github :repo "emacs-lsp/lsp-javascript"))` in your local repo. Then reinstall your `lsp-javascript-typescript` package.

It should work now...😂 BTW, I feel it's counterintuitive for me too...

**Comment by jupl** [comment]
```elisp
        web-beautify
        skewer-mode
        livid-mode
        (lsp-javascript-typescript :requires lsp-mode :location (recipe :fetcher github :repo "emacs-lsp/lsp-javascript"))
        ))

(defun javascript/post-init-add-node-modules-path ()
```

Tried that and reinstalled the package. Still does not work with Tide nor LSP.

**Comment by ztlevi** [comment]
Did you merge the PR?

**Comment by jupl** [comment]
I don't have permission to do that. #10664 does indeed fix this issue.

**Comment by syl20bnr** [comment]
So we need both to merge the PR and use a location recipe to fix the
issue ?

Will this be fixed in MELPA ?

**Comment by syl20bnr** [comment]
I merged the hooks fix.

**Comment by jupl** [comment]
Yeah, it all works.

**Comment by jupl** [comment]
Thank you @syl20bnr @ztlevi. LSP is pretty sweet so far, although I still need to make some tweaks on my part before fully switching over.

**Comment by syl20bnr** [comment]
@jupl @ztlevi Can you remind me why typescript layer depends on node and
javascript layer now ? Is that only because lsp-typescipt is in
lsp-javascript package ?

https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/typescript/layers.el#L12

I would like to add a comment that explain the dependency.

**Comment by jupl** [comment]
@syl20bnr `node` is for the `add-node-modules-path` package, which is disabled by default for security purposes.

**Comment by ztlevi** [comment]
Yep. The javascript dependency is only for the `lsp-javascript-typescript` package. I don't know how to put the same package under different layers. If we can do that, we can just put an identical package under typescript layer.

**Comment by syl20bnr** [comment]
@jupl @ztlevi 

We cannot do that, a package belongs to only one layer, maybe one more
argument in favor of having multiple MELPA packages ;-)

I pushed a bunch of commits to make LSP backends (and tide but not
tested TBH) to work correctly with directory file variables (to be able
to set the backend on a per-project basis)

I encourage you to read the commit messages for the reason we need to do
the setup like that.

Basically this is because we use hack-local-variables-hook to be able to
intercept the value before the file is actually loaded.

**Comment by jupl** [comment]
@syl20bnr I welcome any kind of solution to have multiple packages use a module. The reason that the `node` module existed in the first place was that the `node-add-modules` path is used across `html`, `javascript`, and `typescript` layers. The layers can be loaded independently in any order, plus only one layer can perform the `init-[pkg]` function. [This thread](https://github.com/syl20bnr/spacemacs/pull/8621#discussion_r111666531) discussed how this should be handled. Also [this question was asked here](https://github.com/syl20bnr/spacemacs/issues/8896).

**Comment by syl20bnr** [comment]
@ztlevi Yes we can do that, it should work, until it is in MELPA. I hopd
the package author will submit multiple recipies.

**Comment by syl20bnr** [comment]
@jupl This is the correct way to do it :+1:

**Comment by syl20bnr** [comment]
@jupl @ztlevi I updated the Javascript layer. I wonder if we need the
 default flycheck setup when using the LSP backend:

https://github.com/syl20bnr/spacemacs/blob/develop/layers/+lang/javascript/packages.el#L87-L89

**Comment by ztlevi** [comment]
I'm still using `eslint` as the default linter. I am pretty sure it's the most popular linter for javascript right now since it's more configurable. Especially in my case, I have to turn off it by setting `(setq lsp-ui-flycheck-enable nil)` and make flycheck work properly. What about you @jupl ?
